### ISSUE TITLE
UMA 2.0 RPT Request should be authenticated using the client credentials (not the RqP's OIDC access token)

### ISSUE THREAD

**Issue by randomstuff** [keycloak/keycloak#30779]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authorization-services

### Describe the bug

The UMA 2.0 [RPT request](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html#uma-grant-type) implemented in Keycloak appears to expect the OpenID Connect Access Token as authentication (i.e. in the `Authentication` HTTP header). I don't belive this is what is expected to happen according to the UMA 2.0 specification.

I believe that the intent is to use the **client credentials** for this request (eg. client_id and client_secret with basic authentication). This not very explicit in the specification but the `token_endpoint` is supposed to use client authentication. 

The examples in the specification use a (invalid) Basic authentication:

~~~
POST /token HTTP/1.1
Host: as.example.com
Authorization: Basic jwfLG53^sad$#f
...
grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Auma-ticket
&ticket=016f84e8-f9b9-11e0-bd6f-0021cc6004de
~~~

When a Basic authentication is used on an OAuth token endpoint, the client credentials are used. For example in the OAuth 2.0 (Access Token Request](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.3) of the Code Exchange flow:

> If the client type is confidential or the client was issued client
   credentials (or assigned other authentication requirements), the
   client MUST authenticate with the authorization server as described
   in [Section 3.2.1](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2.1).

The client identity should be sent as a `claim_token` (if any). If no such identity is provided, the authorization server should [redirect the client for an interactive claim gathering](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html#rfc.section.3.3.2).

### Version

25.0

### Regression

- [ ] The issue is a regression

### Expected behavior

The UMA 2.0 RPT Requesr should accept any client authentication.

The UMA 2.0 should not accept a resource owner's or requesting party's access token for authentication.

### Actual behavior

When given a client_id and client_secret in HTTP Basic authentication, the server sends a 403:

~~~json
{"error":"access_denied","error_description":"request_submitted"}
~~~

When passed a requesting party's OpenID Connect Access Token, the request succeeds.

### How to Reproduce?

~~~python
from requests.auth import HTTPBasicAuth, AuthBase
import requests
import sys

AS_URI = "http://localhost:8180/realms/test"
OIDC_CONFIG_ENDPOINT = AS_URI + "/.well-known/openid-configuration"
UMA2_CONFIG_ENDPOINT = AS_URI + "/.well-known/uma2-configuration"

CLIENT_ID = "client"
CLIENT_SECRET = "client-secret"

CLIENT_BASIC_AUTHZ = HTTPBasicAuth(CLIENT_ID, CLIENT_SECRET)

ticket = sys.argv[1]

class BearerAuth(AuthBase):

    def __init__(self, token):
        self.token = token

    def __eq__(self, other):
        return all(
            [
                self.token == getattr(other, "token", None),
            ]
        )

    def __ne__(self, other):
        return not self == other

    def __call__(self, r):
        r.headers["Authorization"] = "Bearer " + self.token
        return r


def json_response(response: requests.Response):
    if response.status_code >= 400:
        print(response.headers)
        print(response.content)
        response.raise_for_status()
    return response.json()

oidc_config = json_response(requests.get(OIDC_CONFIG_ENDPOINT))
uma2_config = json_response(requests.get(UMA2_CONFIG_ENDPOINT))

uma_rpt_response = json_response(
        requests.post(
            uma2_config["token_endpoint"],

            # Unexpected: we make the request using the user's access token.
            # auth=BearerAuth(oidc_state.get("access_token")),

            # I was expecting to use the client credentials
            auth=CLIENT_BASIC_AUTHZ,
            # but I get a 403:
            # {"error":"access_denied","error_description":"request_submitted"}

            data={
                "grant_type": "urn:ietf:params:oauth:grant-type:uma-ticket",
                "ticket": ticket,
            },
        )
    )
~~~

I can provide a minimal test lab if needed.

**Comment by pedroigor** [comment]
You can authenticate using any of the supported client authentication methods.

The message `{"error":"access_denied","error_description":"request_submitted"}` is because the request was not granted, and as you are sending a ticket, a request will be sent to the resource owner to approve the permission on behalf of the subject interested in accessing the resource.

In your case, by only providing the ticket an no access token (e.g.: via header or `subject_token` parameter) are you interested on retrieving the permissions that the client has access to? Wondering what is the use case ...

**Comment by randomstuff** [comment]
I have done some further testing and here is the result of what I observed and what I would expect.

|Authentication header|Claim Token|Expected result|Result|
|--------------------------------|------------------|---|---|
|Client credentials        |-                     |Interactive claims gathering for user authentication | 403 access_denied
|Client credentials        |Bob's ID token |Bob's RPT | Bob's RPT
|Bob's access token      |- |401 Bob's AT should not be accepted for client authentication | Bob's RPT
|Bob's access token      |Bob's ID token |401 Bob's AT should not be accepted for client authentication | Bob's RPT

I believe the `Authorization` HTTP header of the token endpoint is always expected to receive **client authentication** (client credentials). In case 3 and 4, the `Authorization` HTTP is used to receive an **end-user authentication**. I don't believe this is expected behavior for OAuth 2.0.

The token endpoints is expected to receive client authentication as explained in [OAUth 2.0 3.2.1](https://www.rfc-editor.org/rfc/rfc6749#section-3.2.1):

> Confidential clients or other clients issued client credentials MUST
> authenticate with the authorization server as described in
> [Section 2.3](https://www.rfc-editor.org/rfc/rfc6749#section-2.3) when making requests to the token endpoint.

See has well the [Token Exchange spec](https://datatracker.ietf.org/doc/html/rfc8693) request which is similar to the UMA flow but has clearer wording:

> A client requests a security token by making a token request to the authorization server's token endpoint [...]. Client authentication to the authorization server is done using the **normal mechanisms provided by OAuth 2.0** [...]

It references the [JWT profile spec.](https://datatracker.ietf.org/doc/html/rfc7523#section-2.3):

> The JWT MUST contain a "sub" (subject) claim identifying the
> principal that is the subject of the JWT.  Two cases need to be
> differentiated:
>
> A.  For the authorization grant, the subject typically
>            identifies an authorized accessor for which the access token
>            is being requested (i.e., the resource owner or an
>            authorized delegate), but in some cases, may be a
>            pseudonymous identifier or other value denoting an anonymous
>            user.
>
> B.  For client authentication, the subject MUST be the
>            "client_id" of the OAuth client.

Here we are interested in the second case (client authentication).

To summarize these snippets, a JWT passed to the token endpoint using the `Authorization` header is supposed to be about (subject) the client application, not about the end-user. In case (3) and (4), the subject is Bob. I would claim that this usage is not consistent with OAuth.

Moreover in case (3) (and maybe in claim (4)?), the end-users claims are derived from the `Authorization` JWT. As this JWT is supposed to be about the client and not about the end-user, it should not be used to derive end-users claims. End-Users claims are supposed to be transported in the `claim_token`, `pct`, `rpt` claims instead.

**Comment by randomstuff** [comment]
> In your case, by only providing the ticket and no access token (e.g.: via header or subject_token parameter) are you interested on retrieving the permissions that the client has access to? Wondering what is the use case ...

As I have not pushed any client identity, I would expect the authorization server to trigger an interactive claim-gathering process in order to authenticate the RqP (and possibly obtain her consent).

See [3.3.2](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html#claim-redirect) of UMA 2.0:

> [Interactive Claims-Gathering](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html#claim-redirect):
> The client redirects an end-user requesting party to the authorization server's claims interaction endpoint for one or more interactive claims-gathering processes as the authorization server requires. These can include direct interactions, such **as account registration** and **authentication local to the authorization server as an identity provider**, filling out a questionnaire, or asking the user to authorize subsequent collection of claims by interaction or pushing, and persistent storage of such claims (for example, as associated with a PCT). Interactions could also involve further redirection, for example, for **federated (such as social) authentication at a remote identity provider**, and other federated claims gathering.

i.e. if the UMA authorization server could not (or would not) identity the RqP based on pushed claims (in particular when no such pushed claims have been used), it would/should redirect the RqP to its web user interface. This web user interface could be used by the AS to authenticate the RqP. Once authenticated, the authorization server would provide an RPT for the authenticated RqP's to the client application.

Section [1.3](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html#trust-push) explains:

> Further, although both claims pushing and interactive claims gathering are shown, **both might not typically be used in one scenario**.

i.e. interactive claims-gathering would typically be used when no claims have been pushed by the client application.

**Comment by randomstuff** [comment]
See https://github.com/randomstuff/keycloak-uma-test (and test_rs_resource.py) for a test environment.

**Comment by pedroigor** [comment]
As discussed with @randomstuff, the use cases around using a bearer token to authenticate the client are related to proxies where the token from a client application is just forwarded by a RS to grant or deny access to its resource. Without the RS itself authenticating using its credentials.

However, and for certain use cases, we might want to disallow using bearer tokens as a credential and force the client or RS to authenticate using its credentials where the bearer token can then be passed using the `claim_token` or the  `subject_token` parameters.

**Comment by randomstuff** [comment]
According to the specification, the UMA token request is expected to be protected using the client credentials. This prevents the user (and the user's frontend) to directly obtain the RPT (for server-based clients) and therefore prevents the user from abusing the RPT (assuming the server-side client does not send the RPT to the user's frontend). This might be especially interesting in controlled environments which only have manually registered server-side clients.

Because, Keycloak allows obtaining the RPT using the client credentials such as protection is not possible.

### Impact 1, RPT exposed to the user

**Summary:** By obtaining his own access token, the end-user can directly obtain a RPT (associated with the client application) from the authorization server and bypass any control (restriction, rate limiting, anomaly detection) which could be implemented by the server-side client application. The access token can be exposed to the user when using the implicit or hybrid flow.

When using the OAuth code exchange or the UMA flow with confidential server-side clients, the access token / RPT is not actually exposed to the user / client frontend (see the [BFF pattern](https://www.ietf.org/archive/id/draft-ietf-oauth browser-based-apps-15.html#name-backend-for-frontend-bff)) unless the client backend actually passes the RPT to the client frontend (see the [Token Mediating pattern](https://www.ietf.org/archive/id/draft-ietf-oauth-browser-based-apps-15.html#name-token-mediating-backend)):

* This might limit the attacks a malicious user can take on the resource server as the user can only take actions on the resource server through the client backend.
* This might prevent the malicious user from seeing some data which would not be exposed by the backend server to the user frontend.



The OAuth 2.0 spec [mentions the idea that one might want to avoid exposing the tokens to the user's frontend](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2):

> [Implicit Grant] Because the access token is encoded into the redirection URI, it <mark>may be exposed to the resource owner</mark> and other applications residing on the same device.
>
> [web application profile] The client credentials as well as any access token issued to the client are stored on the web server and are <mark>not exposed to or accessible by the resource owner</mark>.

The [OAuth 2.0 for Browser-Based Apps draft](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps) explains that one benefit of the [BFF pattenrs](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#name-backend-for-frontend-bff) is that it does not expose the tokens to the user / front-end:

> The BFF counters the first two payloads by <mark>not exposing any tokens to the browser-based application</mark>. Even when the attacker gains full control over the JavaScript application, there are simply no tokens to be stolen.

Moreover the BFF can implement rate limiting, anomaly detection:

> If <mark>a high-security application would prefer to implement anomaly detection or rate limiting</mark>, such a BFF would be the ideal place to do so. Such restrictions can further help to mitigate the consequences of client hijacking.

***

Let's assume that in your realm authorization server you have:

* RS, a UMA resource server.
* APP, an server-based application (using the BFF pattern) which supports the OpenID Connect hybrid flow and acts as a a UMA client for RS;
* a malicious user.

We might want to use the BFF pattern in order to avoid exposing the RPT to the user: the malicious user would not be able to make direct calls to the RS but would be restricted by the logic of the server-side BFF. This would make it harder for the malicious attacker to conduct an attack on the RS.



However, in this scenario, the malicious user could use the fact that client authentication is not required by Keycloak (only the user OpenID Connect access token) for obtaining the RPT:

1. the malicious user authenticates himself on the authorization server using the OpenID Connect hybrid flow;
2. the malicious user obtains an OpenID Connect access token;
3. the malicious user makes a direct request to the authorization server using his OpenID Connect access token in order to obtain a RPT token for accessing some protected resources on RS;
4. the malicious attacker can now make direct calls to the RS using this RPT.

### Impact 2, malicious resource server obtaining a RPT for another resource server

Moreover, as any user access token can be used to obtain an RPT, the resource server 1 can use an access token it received from a client application (in particular a RPT) in order to obtain a RPT for accessing the resources on another resource server 2 on the user's behalf:

* This might not be highly problematic if all your resource servers are under your control. Even in this case, this could be used for lateral movement: an attacker taking obtaining control of RS1 could use this position to obtain RPTs for RS2.
* This might be more problematic if some resource servers are not under your control.

**Comment by randomstuff** [comment]
Here is a summary of different accepted combinations for UMA token requests (`…:grant-type:uma-ticket`):

|Authentication     |claim_token       |
|-------------------|------------------|
|client credentials |user ID token     |
|client credentials |user access token |
|client credentials |user refresh token|
|user access token  |none              |

On all cases, Keycloak does not check that the permission ticket, the client credentials and the claim token are associated with the same resource server.

For example, RS1 can use a user token (for example a RPT or an ID token), it received from the user in order to obtain a RPT for resources on RS2 (on the user's behalf).
